crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -6,68 +6,54 @@ use crate::cst::{NamedNode, Node, RuleNode};
 use crate::kinds::{FieldName, RuleKind, TokenKind};
 use crate::text_index::{TextIndex, TextRange};
 
-/// A [`PathNode`] that points to a [`RuleNode`].
+/// A node in the ancestor path of a [`Cursor`].
 #[derive(Clone, Debug, PartialEq, Eq)]
-struct PathRuleNode {
+struct PathAncestor {
+    parent: Option<Rc<PathAncestor>>,
     rule_node: Rc<RuleNode>,
     child_number: usize,
     text_offset: TextIndex,
 }
 
-impl PathRuleNode {
-    fn into_path_node(self) -> PathNode {
-        PathNode {
-            node: Node::Rule(self.rule_node),
-            child_number: self.child_number,
-            text_offset: self.text_offset,
-        }
-    }
-}
-
-/// A pointer to a [`Node`] in a CST, used by the [`Cursor`] to implement the traversal.
+/// A cursor that can traverse a CST.
+///
+/// Nodes are visited in a DFS pre-order traversal.
 #[derive(Clone, Debug, PartialEq, Eq)]
-struct PathNode {
+pub struct Cursor {
+    /// The parent path of this cursor
+    parent: Option<Rc<PathAncestor>>,
     /// The node the cursor is currently pointing to.
     node: Node,
     /// The index of the current child node in the parent's children.
     // Required to go to the next/previous sibling.
     child_number: usize,
     /// Text offset that corresponds to the beginning of the currently pointed to node.
     text_offset: TextIndex,
+    /// Whether the cursor is completed, i.e. at the root node as a result of traversal (or when `complete`d).
+    /// If `true`, the cursor cannot be moved.
+    is_completed: bool,
 }
 
-impl PathNode {
-    fn text_range(&self) -> TextRange {
-        let start = self.text_offset;
-        let end = start + self.node.text_len();
-        start..end
-    }
-
-    fn to_path_rule_node(&self) -> Option<PathRuleNode> {
+impl Cursor {
+    fn as_ancestor_node(&self) -> Option<Rc<PathAncestor>> {
         if let Node::Rule(rule_node) = &self.node {
-            Some(PathRuleNode {
+            Some(Rc::new(PathAncestor {
+                parent: self.parent.clone(),
                 rule_node: rule_node.clone(),
                 child_number: self.child_number,
                 text_offset: self.text_offset,
-            })
+            }))
         } else {
             None
         }
     }
-}
 
-/// A cursor that can traverse a CST.
-///
-/// Nodes are visited in a DFS pre-order traversal.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Cursor {
-    /// The list of ancestor rule nodes that the `current` node is a part of.
-    path: Vec<PathRuleNode>,
-    /// The node the cursor is currently pointing to.
-    current: PathNode,
-    /// Whether the cursor is completed, i.e. at the root node as a result of traversal (or when `complete`d).
-    /// If `true`, the cursor cannot be moved.
-    is_completed: bool,
+    fn set_from_ancestor_node(&mut self, ancestor: &Rc<PathAncestor>) {
```

Comment:
I wonder if this should `self.is_completed = false;` as well, in case the cursor is already completed?

Reply:
This private method is only called when `is_completed == true`.